examples/motoko_examples/basic-dao/src/basic_dao.ts

```diff
@@ -0,0 +1,374 @@
+  // return Array.from(accounts, ([owner, tokens]) => ({ owner, tokens }));
+  const arr: Account[] = [];
+  ic.stableStorage<StableStorage>().accounts.forEach((tokens, owner) => {
+    arr.push({
+      owner,
+      tokens,
+    });
+  });
+
+  return arr;
+}
+
+// / Submit a proposal
+// /
+// / A proposal contains a canister ID, method name and method args. If enough users
+// / vote "yes" on the proposal, the given method will be called with the given method
+// / args on the given canister.
+export function submitProposal(payload: ProposalPayload): Update<Res> {
+  let caller = ic.caller();
+  let result = deductProposalSubmissionDeposit(caller);
+  if ('ok' in result) {
+    let proposal_id: nat = ic.stableStorage<StableStorage>().nextProposalId;
+    ic.stableStorage<StableStorage>().nextProposalId += 1n;
+
+    let proposal = {
+      id: proposal_id,
+      timestamp: ic.time(),
+      proposer: caller,
+      payload,
+      state: { Open: null },
+      votes_yes: zeroToken,
+      votes_no: zeroToken,
+      voters: [],
+    };
+    proposalPut(proposal_id, proposal);
+    return { ok: proposal_id };
+  } else {
+    return { err: result.err };
+  }
+}
+
+/// Return the proposal with the given ID, if one exists
+export function getProposal(proposal_id: nat): Query<Opt<Proposal>> {
+  return proposalGet(proposal_id);
+}
+
+/// Return the list of all proposals
+export function listProposals(): Query<Proposal[]> {
+  // this is causing errors
+  //return Array.from(proposals, ([_, proposal]) => proposal);
+  const arr: Proposal[] = [];
+  ic.stableStorage<StableStorage>().proposals.forEach((proposal) => {
+    arr.push(proposal);
+  });
+
+  return arr;
+}
+
+type VoteResult = Variant<{
+  ok?: ProposalState;
+  err?: string;
+}>;
+// Vote on an open proposal
+export function vote(args: VoteArgs): Update<VoteResult> {
+  let caller = ic.caller();
+  let proposal = proposalGet(args.proposal_id);
+  if (proposal === null) {
+    return { err: `No proposal with ID  ${args.proposal_id} exists` };
+  }
+  // I am keeping the if statemnts here even though they are not needed for readability
+  // as this whole block should only be excuted if there's a proposal
+  if (proposal) {
+    let state = proposal.state;
+    if (!('Open' in state)) {
+      return {
+        err: `Proposal ${args.proposal_id} is not open for voting"`,
+      };
+    }
+    let account = accountGet(caller);
+    if (account === null) {
+      return { err: 'Caller does not have any tokens to vote with' };
+    }
+    if (account) {
+      if (proposal.voters.some((e) => e === caller)) {
+        return { err: 'Already voted' };
+      }
+
+      let votes_yes = proposal.votes_yes.amount_e8s;
+      let votes_no = proposal.votes_no.amount_e8s;
+      if ('Yes' in args.vote) {
+        votes_yes += account.amount_e8s;
+      }
+      if ('No' in args.vote) {
+        votes_no += account.amount_e8s;
+      }
+      proposal.voters.push(caller);
```

Comment:
I think we should create a new array here and do the mutation at the end. Please check what the List.push in Motoko does. If it pushes into an array creating a new copy, then I think we should do that (as we want to manage our mutations and do them all at once if possible).

If that's not what Motoko does, then I suppose we can leave it.

I'm trying to balance best practices with just following what the Motoko code does. I think the Motoko code could be improved in many ways but I want the Azle Motoko examples to be mirrors of the Motoko examples.

Reply:
In Motoko they are always creating a new array instead of mutating, I also prefer immutability in general.